x-pack/packages/ml/response_stream/client/use_fetch_stream.ts

```diff
@@ -75,11 +76,41 @@ export function useFetchStream<B extends object, R extends Reducer<any, any>>(
     ? customReducer
     : ({ reducer: stringReducer, initialState: '' } as FetchStreamCustomReducer<R>);
 
-  const [data, dispatch] = useReducer(
-    reducerWithFallback.reducer,
-    reducerWithFallback.initialState
-  );
-  const dataThrottled = useThrottle(data, 100);
+  // We used `useReducer` in previous iterations of this hook, but it caused
+  // a lot of unnecessary re-renders even in combination with `useThrottle`.
+  // We're now using `dataRef` to allow updates outside of the render cycle.
+  // When the stream is running, we'll update `data` with the `dataRef` value
+  // periodically.
+  const [data, setData] = useState(reducerWithFallback.initialState);
+  const dataRef = useRef(reducerWithFallback.initialState);
+
+  // This effect is used to throttle the data updates while the stream is running.
+  // It will update the `data` state with the current `dataRef` value every 100ms.
+  useEffect(() => {
+    // We cannot check against `isRunning` in the `setTimeout` callback, because
+    // we would check against a stale value. Instead, we use a mutable
+    // object to keep track of the current state of the effect.
+    const effectState = { isActive: true };
+
+    if (isRunning) {
+      setData(dataRef.current);
+
+      function updateData() {
+        setTimeout(() => {
+          setData(dataRef.current);
+          if (effectState.isActive) {
+            updateData();
+          }
+        }, DATA_THROTTLE_MS);
```

Comment:
Should the consumer of this `useFetchStream` hook also have the ability to pass in the duration they want to throttle? Not necessary for this PR but would be nice to have in a follow up.